features/26-mfa/implementation.md

```diff
@@ -0,0 +1,75 @@
+# Database Tables
+
+## _auth_authenticator
+
+|Column|Type|
+|------|----|
+|id|TEXT|
+|type|TEXT; `totp`, `oob`, `recovery_code`, `bearer_token`|
+|user_id|TEXT; foreign key to `_core_user.id`|
+
+## _auth_authenticator_totp
+
+|Column|Type|
+|------|----|
+|id|TEXT; foreign key to `_auth_authenticator.id`|
+|secret|TEXT|
+|activated|BOOLEAN|
+|created_at|timestamp|
+|activated_at|timestamp|
+|display_name|TEXT|
+
+- `activated` default to `false`.
+- Only `activated = true` is considered an authenticator.
+
+## _auth_authenticator_oob
+
+|Column|Type|
+|------|----|
+|id|TEXT; foreign key to `_auth_authenticator.id`|
+|channel|TEXT; `sms` or `email`|
+|phone|TEXT|
+|email|TEXT|
+|activated|BOOLEAN|
+|created_at|timestamp|
+|activated_at|timestamp|
+
+- `activated` default to `false`.
+- Only `activated = true` is considered an authenticator.
+
+## _auth_authenticator_oob_code
+
+|Column|Type|
+|------|----|
+|id|TEXT|
+|user_id|TEXT; foreign key to `_core_user.id`|
+|code|TEXT|
+|created_at|timestamp|
+|expire_at|timestamp|
+
+- A OOB code is generated when trigger is called.
+- A OOB code is valid if it does not expire yet.
+- If a valid OOB code is found, no new code is generated. The valid code is delivered again.
```

Comment:
I'm not sure if this is common practice, but my intuition is that a new code is generated each time the user request a SMS code.

Reply:
From my experience, the OOB code of other online services seems to be the same if triggered multiple times within a short period of time.

Reply:
Okay then.